src/vm/gdbjit.cpp

```diff
@@ -0,0 +1,1075 @@
+
+DwarfLineNumHeader LineNumHeader = {
+    0, 2, 0, 1, 1, DWARF_LINE_BASE, DWARF_LINE_RANGE, DWARF_OPCODE_BASE, {0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1}
+};
+
+/* Static data for .debug_info */
+struct __attribute__((packed)) DebugInfo
+{
+    uint8_t m_cu_abbrev;
+    uint32_t m_prod_off;
+    uint16_t m_lang;
+    uint32_t m_cu_name;
+    uint32_t m_line_num;
+    
+    uint8_t m_sub_abbrev;
+    uint32_t m_sub_name;
+    uint8_t m_file, m_line;
+    uint32_t m_sub_type;
+    
+    uint8_t m_type_abbrev;
+    uint32_t m_type_name;
+    uint8_t m_encoding;
+    uint8_t m_byte_size;
+} debugInfo = {
+    1, 0, DW_LANG_C89, 0, 0,
+    2, 0, 1, 1, 37,
+    3, 0, DW_ATE_signed, 4
+};
+
+/* Create ELF/DWARF debug info for jitted method */
+void NotifyGdb::MethodCompiled(MethodDesc* MethodDescPtr)
+{
+    PCODE pCode = MethodDescPtr->GetNativeCode();
+
+    if (pCode == NULL)
+        return;
+    unsigned int symInfoLen = 0;
+    SymbolsInfo* symInfo = nullptr;
+
+    /* Get method name & size of jitted code */
+    LPCUTF8 methodName = MethodDescPtr->GetName();
+    EECodeInfo codeInfo(pCode);
+    TADDR codeSize = codeInfo.GetCodeManager()->GetFunctionSize(codeInfo.GetGCInfo());
+    
+#ifdef _TARGET_ARM_    
+    pCode &= ~1; // clear thumb flag for debug info
+#endif    
+
+    /* Get module name */
+    const Module* mod = MethodDescPtr->GetMethodTable()->GetModule();
+    SString modName = mod->GetFile()->GetPath();
+    StackScratchBuffer scratch;
+    const char* szModName = modName.GetUTF8(scratch);
+    const char *szModulePath, *szModuleFile;
+    
+    SplitPathname(szModName, szModulePath, szModuleFile);
+    
+    /* Get debug info for method from portable PDB */
+    HRESULT hr = GetDebugInfoFromPDB(MethodDescPtr, &symInfo, symInfoLen);
+    if (FAILED(hr) || symInfoLen == 0)
+    {
+        return;
+    }
+
+    MemBuf elfHeader, sectHeaders, sectStr, dbgInfo, dbgAbbrev, dbgPubname, dbgPubType, dbgLine, dbgStr, elfFile;
+
+    /* Build .debug_abbrev section */
+    if (!BuildDebugAbbrev(dbgAbbrev))
+    {
+        delete[] symInfo;
+        return;
+    }
+    
+    /* Build .debug_line section */
+    if (!BuildLineTable(dbgLine, pCode, symInfo, symInfoLen))
+    {
+        delete[] symInfo;
+        delete[] dbgAbbrev.MemPtr;
+        return;
+    }
+    
+    DebugStrings[1] = szModuleFile;
+    DebugStrings[3] = methodName;
+    
+    /* Build .debug_str section */
+    if (!BuildDebugStrings(dbgStr))
+    {
+        delete[] symInfo;
+        delete[] dbgAbbrev.MemPtr;
+        delete[] dbgLine.MemPtr;
+        return;
+    }
+    
+    /* Build .debug_info section */
+    if (!BuildDebugInfo(dbgInfo))
+    {
+        delete[] symInfo;
+        delete[] dbgAbbrev.MemPtr;
+        delete[] dbgLine.MemPtr;
+        delete[] dbgStr.MemPtr;
+        return;
+    }
+    
+    /* Build .debug_pubname section */
+    if (!BuildDebugPub(dbgPubname, methodName, dbgInfo.MemSize, 26))
+    {
+        delete[] symInfo;
```

Comment:
I recommend using NewArrayHolder wrapper for all of these arrays. Then you don't have to do any cleanup explicitly and repeat it over and over, the destructor of the NewArrayHolder would do that for you. When / if you need to keep a pointer over the lifetime of the holder, you just call SuppressRelease on the holder.